### PR TITLE
WIP: Initial work to drop Simplecov and Simplecov HTML as dependencies

### PR DESCRIPTION
**Pull request by danmayer**
This is a start to build out full support for a mountable Rack UI which can support additional features beyond the original static reports.

* Showing the time range the coverage is valid for
* Ability to filter coverage to a specific time range
* Improved UI and navigation
* Gem coverage usage summaries
* etc

This pulls the initial views, assets (JS, CSS) from Simple-Cov HTML as well as some files to do calculations on coverage data into the project. Initially with little changes.

### CODE REVIEW

File: lib/coverband/reporters/html_report.rb [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+          Coverband::Utils::HTMLFormatter.new(filtered_report_files).format_html!
+        else
+          Coverband::Utils::HTMLFormatter.new(filtered_report_files).format!
+          if open_report
+            `open #{Coverband.configuration.root}/coverage/index.html`
+          else
+            Coverband.configuration.logger.info 'report is ready and viewable: open coverage/index.html'
+          end
+
+          Coverband::Utils::S3Report.instance.persist! if Coverband.configuration.s3_bucket
```
[the remark this answers is not in the document]
Comment (danmayer) [reply lib/coverband/reporters/html_report.rb]: Yeah that was a question I had as well... Is there any good use cases for generating the reporting and writing to S3... 

What I came up with:
* If you didn't want to use the Rack-mountable endpoints at all
* Archiving... if you write to S3 with the audit function of S3 it will have all your coverage history (at least every version you uploaded)
* CI based workflows, where CI is just generating and making reports like this visible (what I used to do before the rack mountable endpoint)

I can't decide how many folks want these types of features vs mounting the new web piece.

**Comment by danmayer** [comment]
Good call on a capybara or any full stack tests which also pulls the report. It would also be easy to add a web endpoint that returns the data as JSON, which could make for really easy test automation, in terms of parsing and checking the results.

**Comment by kbaum** [comment]
Love the idea of a json api but verifying the html report feels more like an end to end feature test.

File: lib/coverband/reporters/html_report.rb (hunk 1 above)
[the remark this answers is not in the document]
Comment (kbaum) [reply lib/coverband/reporters/html_report.rb]: All good points. Difficult to get a handle on what features people need. Almost need coverband for coverband 🤔 .

How about we deprecate the s3 report for this release? Could put a log message and a note in the README.

File: lib/coverband/reporters/html_report.rb [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+
+        # list all files, even if not tracked by Coverband (0% coverage)
+        tracked_glob = "#{current_root}/{app,lib,config}/**/*.{rb}"
+        report_files = Coverband::Utils::Result.add_not_loaded_files(scov_style_report, tracked_glob)
+        # apply coverband filters
+        filtered_report_files = {}
+        report_files.each_pair do |file, data|
+          next if Coverband.configuration.ignore.any? { |i| file.match(i) }
+          filtered_report_files[file] = data
+        end
```
Comment (kbaum) [inline lib/coverband/reporters/html_report.rb]: Good use of each_with_object? Just a style thing... i dunno what's better.
```ruby
filtered_report_files = report_files.each_with_object({}) do |(file, data), hash|
  hash[file] = data unless Coverband.configuration.ignore.any? { |i| file.match(i) }
end
```

**Comment by danmayer** [comment]
still looking to add tests and clean this up a bit more I have been sick and a bit busy with work recently, but should get back to it soon.

**Comment by danmayer** [comment]
ok made a bit more progress pulled in all tests from one class and I think it shows a pretty good pattern to get the rest added.

**Comment by danmayer** [comment]
almost done pulling in all the related tests that was a lot harder / more time to consume than I would have guessed... hopefully will have that done in next day or so... down to a few failing edge cases I need to understand but the majority of tests passing now.

**Comment by danmayer** [comment]
ok all tests passing the only remaining failure is the full stack test that you are working and changing @kbaum I think one more pass on clean up by me and this might be ready for review and merge.

**Comment by kbaum** [comment]
@danmayer Struggling a bit with rails end to end test sporadically failing. Trying to pin down why. Hopefully get some time to work on it more today:

https://github.com/danmayer/coverband/pull/150/files

**Comment by danmayer** [comment]
ok, @kbaum take a look I am thinking it makes sense to merge this now as all the tests and functionality are there... I want to do another pass cleaning up the web report functionality before the release... I am thinking I will get al the rest of the HTML into templates and try to add some basic styles on the rack web index as well as make all of the features a bit more self-documented.
